pkgs/tools/backup/borgbackup/default.nix

```
@@ -7,21 +7,37 @@
 , openssh
 , openssl
 , python3
+, xxHash
 , zstd
 , installShellFiles
 , nixosTests
+, fetchpatch
 }:
 
 python3.pkgs.buildPythonApplication rec {
   pname = "borgbackup";
-  version = "1.2.2";
+  version = "1.2.3";
   format = "pyproject";
 
   src = python3.pkgs.fetchPypi {
     inherit pname version;
-    sha256 = "sha256-1zBodEPxvrYCsdcrrjYxj2+WVIGPzcUEWFQOxXnlcmA=";
+    sha256 = "sha256-4yQY+GM8lvqWgTUqVutjuY4pQgNHLBFKUkJwnTaWZ4U=";
   };
 
+  patches = [
+    # Revert faulty test change for HashIndexSizeTestCase.test_size_on_disk_accurate
+    (fetchpatch {
+      url = "https://github.com/borgbackup/borg/commit/05997e2867b2abb826abf9e9a564e288302bce8c.patch";
+      hash = "sha256-NORv1XAmM39hXOz4AxbByzydZ5Zn3NbiWJeZHmLb0hI=";
+      revert = true;
+    })
+    (fetchpatch {
+      url = "https://github.com/borgbackup/borg/commit/cc0ad321dc32b78ce2f2f625ae91040fddf3fd8c.patch";
+      hash = "sha256-kM9TfjMJkmGr3XZGB85HpcwuBJV8Y/vHEGnlDMUhKso=";
+      revert = true;
+    })
```

i.e. to revert this change using the commit from the (1.2-maint) branch rather than the similar one from 2.0.0b4

```suggestion
    (fetchpatch {
      url = "https://github.com/borgbackup/borg/commit/8ddfd94f0904449175151faed67f9640710b6454.patch";
      hash = "sha256-+6ELPrvCTb21my6WanpqbPl2Sl1oxhb/z1zpo1KbMhc=";
      revert = true;
    })
```

---

but I don't see any pytest failure to begin with, even with that change left in (i.e. no patching)

---

The failure is a race condition. The actual fix would be to to call `flush()` and `fsync()`.

---

Ah, and so the key difference is in 
https://github.com/borgbackup/borg/blob/c29d4a096bba4018e14208b54fe24bdc886f4f3d/src/borg/hashindex.pyx#L123-L128

- given a str, it's used as a pathname to `with open(path, 'wb') as fd:`, and the file is closed before ChunkIndex.write returns. Given a file object, the write goes to the existing file object, which may not have flushed to the filesystem yet.
I don't think `fsync` should relevant - this doesn't care if the data is safe at rest on the physical medium. But it definitely matters whether or not cpython has actually made the relevant `write` syscall yet, or whether it's just spooling in-process.

---

Ah, it will prefer `st_blksize` over `DEFAULT_BUFFER_SIZE` if the filesystem where the tempfile lives answers this stat: https://github.com/python/cpython/blob/024ac542d738f56b36bdeb3517a10e93da5acab9/Lib/_pyio.py#L247-L255 or (on most platforms) similar C code: https://github.com/python/cpython/blob/85dd6cb6df996b1197266d1a50ecc9187a91e481/Modules/_io/bufferedio.c

So, dooing a little more tracing
```py
import io
import tempfile
print(io.DEFAULT_BUFFER_SIZE)

with tempfile.NamedTemporaryFile() as file:
    pass
```

`strace --trace=file --no-abbrev  python temp_test.py`

shows 
```
openat(AT_FDCWD, "/run/user/1000/tmp486dtl2p", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW|O_CLOEXEC, 0600) = 3
newfstatat(3, "", {st_dev=makedev(0, 0x3b), st_ino=1329, st_mode=S_IFREG|0600, st_nlink=1, st_uid=1000, st_gid=100, st_blksize=4096, st_blocks=0, st_size=0, st_atime=1673231853 /* 2023-01-08T20:37:33.136371805-0600 */, st_atime_nsec=136371805, st_mtime=1673231853 /* 2023-01-08T20:37:33.136371805-0600 */, st_mtime_nsec=136371805, st_ctime=1673231853 /* 2023-01-08T20:37:33.136371805-0600 */, st_ctime_nsec=136371805}, AT_EMPTY_PATH) = 0
unlink("/run/user/1000/tmp486dtl2p")    = 0
```

On my btrfs, it got `st_blksize=4096`, so even less than the default.

I assume in whatever environment you're seeing failures, it must be considerably larger (at least > 90350 for python to not have written through anything at all), since BufferedWriter should flush through:

https://docs.python.org/3/library/io.html#io.BufferedWriter
> - when the buffer gets too small for all pending data
> -  when [flush()](https://docs.python.org/3/library/io.html#io.BufferedWriter.flush) is called;
> - when a seek() is requested (for [BufferedRandom](https://docs.python.org/3/library/io.html#io.BufferedRandom) objects);
> - when the [BufferedWriter](https://docs.python.org/3/library/io.html#io.BufferedWriter) object is closed or destroyed.

---

/run is a tmpfs, so, by process of elimination - you are probably using a kernel with different transparent hugepage settings: https://www.kernel.org/doc/html/v5.9/admin-guide/mm/transhuge.html#hugepages-in-tmpfs-shmem, as tmpfs uses the memory page size for st_blksize. Which would be one of the few things nix hasn't locked down for reproducibility...

---

I can reproduce the issue on one of my remote builders, that does not use `tmpfs` for `/tmp`, because nix dumps pretty large things into there sometimes and doesn't always clean up after itself.

On my desktop however I have a tmpfs and the test works there.

---

tmpfs without hugetlb (the default) has 4kiB pages, as I'm seeing and that "works" (though there should indeed be a flush() in the code, that's a legit bug in the test now). I was just continuing to dig at the non-determinism, since that's also unexpected in an environment as tightly-controlled as nix, and I didn't want to be the one trying to open an upsteam ticket for a problem I didn't understand and couldn't reproduce :-)

It seems like environment you're seeing failures on (apparently zfs) must have `st_blksize` > 90350 for python not to have written anything at all. https://docs.oracle.com/cd/E23824_01/html/821-1448/gazss.html says "The default block size for volumes is 8 KB. Any power of 2 from 512 bytes to 128 KB is valid." but it seems to be something set at volume creation time.

I guess you could check with strace and we'd know for sure, but finding out cpython's behavior is based on st_blksize is probably enough to know where something the derivation doesn't control could come from.

---

Opened https://github.com/borgbackup/borg/issues/7250 and speculated it is related to ZFS' 5 second commit interval, but since you dug deeper you might as well chime in.